client/my-sites/plans-features-main/index.tsx

```diff
@@ -421,6 +362,47 @@ const PlansFeaturesMain = ( {
 
 	const eligibleForFreeHostingTrial = useSelector( isUserEligibleForFreeHostingTrial );
 
+	const planActionCallback = ( planSlug: PlanSlug ): boolean => {
+		if (
+			sitePlanSlug &&
+			isFreePlan( sitePlanSlug ) &&
+			domainFromHomeUpsellFlow &&
+			intentFromProps !== 'plans-p2'
+		) {
+			showDomainUpsellDialog();
+			return true;
+		}
+
+		setLastClickedPlan( planSlug );
+
+		const displayedModal = resolveModal( planSlug );
+		if ( displayedModal ) {
+			setIsModalOpen( true );
+			return true;
+		}
+
+		return false;
+	};
```

Comment:
I've been mulling this over and still don't have a great answer for the naming 🤔 

It's a bit challenging because the function is serving two very distinct responsibilities. It short circuits normal cta logic _and_ it also shows some modal or upsell.

Perhaps `exitEarlyAndShowModal`? Thinking this over a bit more

Reply:
Yeah, naming this will be hard. Something tells me we need to reorganize things a little (wrap differently) and avoid passing it through as a prop to be called from the other hook. We don't need to "centralize" everything.

I can revisit/try something tomorrow if nothing else comes to mind.

Reply:
Hey Christos unfortunately I didn't get to this today, but will start focusing on this plan action callback first thing tomorrow